File: go/indicators/patrickmulloy/doubleexponentialmovingaverage/doubleexponentialmovingaverage.go
package doubleexponentialmovingaverage

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
)

// DoubleExponentialMovingAverage computes the Double Exponential Moving Average (DEMA),
// a smoothing indicator with less lag than a straight exponential moving average.
//
// The DEMA was developed by Patrick G. Mulloy and is described in two articles:
//
//	❶ Technical Analysis of Stocks & Commodities v.12:1 (11-19), Smoothing Data With Faster Moving Averages.
//	❷ Technical Analysis of Stocks & Commodities v.12:2 (72-80), Smoothing Data With Less Lag.
//
// The calculation is as follows:
//
//	EMA¹ᵢ = EMA(Pᵢ) = αPᵢ + (1-α)EMA¹ᵢ₋₁ = EMA¹ᵢ₋₁ + α(Pᵢ - EMA¹ᵢ₋₁), 0 < α ≤ 1
//	EMA²ᵢ = EMA(EMA¹ᵢ) = αEMA¹ᵢ + (1-α)EMA²ᵢ₋₁ = EMA²ᵢ₋₁ + α(EMA¹ᵢ - EMA²ᵢ₋₁), 0 < α ≤ 1
//	DEMAᵢ = 2 EMA¹ᵢ - EMA²ᵢ
//
// The very first EMA value (the seed for subsequent values) is calculated differently.
// This implementation allows for two algorithms for this seed.
//
//	❶ Use a simple average of the first 'period'. This is the most widely documented approach.
//	❷ Use first sample value as a seed. This is used in Metastock.
type DoubleExponentialMovingAverage struct {
	mu sync.RWMutex
	core.LineIndicator
	smoothingFactor float64
	sum             float64
	ema1            float64
	ema2            float64
	length          int
	length2         int
	count           int
	firstIsAverage  bool
	primed          bool
}

// NewDoubleExponentialMovingAverageLength returns an instnce of the indicator
// created using supplied parameters based on length.
func NewDoubleExponentialMovingAverageLength(
	p *DoubleExponentialMovingAverageLengthParams,
) (*DoubleExponentialMovingAverage, error) {
	return newDoubleExponentialMovingAverage(p.Length, math.NaN(), p.FirstIsAverage,
		p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

// NewDoubleExponentialMovingAverageSmoothingFactor returns an instnce of the indicator
// created using supplied parameters based on smoothing factor.
func NewDoubleExponentialMovingAverageSmoothingFactor(
	p *DoubleExponentialMovingAverageSmoothingFactorParams,
) (*DoubleExponentialMovingAverage, error) {
	return newDoubleExponentialMovingAverage(0, p.SmoothingFactor, p.FirstIsAverage,
		p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

//nolint:funlen,cyclop
func newDoubleExponentialMovingAverage(length int, alpha float64, firstIsAverage bool,
	bc entities.BarComponent, qc entities.QuoteComponent, tc entities.TradeComponent,
) (*DoubleExponentialMovingAverage, error) {
	const (
		invalid = "invalid double exponential moving average parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtnl   = "dema(%d%s)"
		fmtna   = "dema(%d, %.8f%s)"
		minlen  = 1
		two     = 2.
		epsilon = 0.00000001
	)

	var (
		mnemonic  string
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	// Resolve defaults for component functions.
	// A zero value means "use default, don't show in mnemonic".
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	if math.IsNaN(alpha) {
		if length < minlen {
			return nil, fmt.Errorf(fmts, invalid, "length should be positive")
		}

		alpha = two / float64(1+length)
		mnemonic = fmt.Sprintf(fmtnl, length, core.ComponentTripleMnemonic(bc, qc, tc))
	} else {
		if alpha < 0. || alpha > 1. {
			return nil, fmt.Errorf(fmts, invalid, "smoothing factor should be in range [0, 1]")
		}

		if alpha < epsilon {
			alpha = epsilon
		}

		length = int(math.Round(two/alpha)) - 1
		mnemonic = fmt.Sprintf(fmtna, length, alpha, core.ComponentTripleMnemonic(bc, qc, tc))
	}

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	desc := "Double exponential moving average " + mnemonic

	dema := &DoubleExponentialMovingAverage{
		smoothingFactor: alpha,
		length:          length,
		length2:         int(two)*length - 1,
		firstIsAverage:  firstIsAverage,
	}

	dema.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, dema.Update)

	return dema, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *DoubleExponentialMovingAverage) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
// It always has a single scalar output -- the calculated value of the double exponential moving average.
func (s *DoubleExponentialMovingAverage) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.DoubleExponentialMovingAverage,
		s.LineIndicator.Mnemonic,
		s.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: s.LineIndicator.Mnemonic, Description: s.LineIndicator.Description},
		},
	)
}

// Update updates the value of the indicator given the next sample.
func (s *DoubleExponentialMovingAverage) Update(sample float64) float64 { //nolint:cyclop
	const two = 2.

	if math.IsNaN(sample) {
		return sample
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	if s.primed {
		v1 := s.ema1
		v2 := s.ema2
		sf := s.smoothingFactor
		v1 += (sample - v1) * sf
		v2 += (v1 - v2) * sf
		s.ema1 = v1
		s.ema2 = v2

		return two*v1 - v2
	}

	s.count++
	if s.firstIsAverage { //nolint:nestif
		if s.count == 1 {
			s.sum = sample
		} else if s.length >= s.count {
			s.sum += sample
			if s.length == s.count {
				s.ema1 = s.sum / float64(s.length)
				s.sum = s.ema1
			}
		} else { // if s.length2 >= s.count {
			s.ema1 += (sample - s.ema1) * s.smoothingFactor
			s.sum += s.ema1

			if s.length2 == s.count {
				s.primed = true
				s.ema2 = s.sum / float64(s.length)

				return two*s.ema1 - s.ema2
			}
		}
	} else { // Metastock
		if s.count == 1 {
			s.ema1 = sample
		} else if s.length >= s.count {
			s.ema1 += (sample - s.ema1) * s.smoothingFactor
			if s.length == s.count {
				s.ema2 = s.ema1
			}
		} else { // if s.length2 >= s.count {
			s.ema1 += (sample - s.ema1) * s.smoothingFactor
			s.ema2 += (s.ema1 - s.ema2) * s.smoothingFactor

			if s.length2 == s.count {
				s.primed = true

				return two*s.ema1 - s.ema2
			}
		}
	}

	return math.NaN()
}

File: py/indicators/patrick_mulloy/double_exponential_moving_average/double_exponential_moving_average.py
"""Double exponential moving average indicator."""

import math

from ...core.indicator import Indicator
from ...core.line_indicator import LineIndicator
from ...core.metadata import Metadata
from ...core.build_metadata import build_metadata, OutputText
from ...core.identifier import Identifier
from ...core.component_triple_mnemonic import component_triple_mnemonic
from ...core.output import Output
from ....entities.bar import Bar
from ....entities.quote import Quote
from ....entities.trade import Trade
from ....entities.scalar import Scalar
from ....entities.bar_component import BarComponent, DEFAULT_BAR_COMPONENT, bar_component_value
from ....entities.quote_component import QuoteComponent, DEFAULT_QUOTE_COMPONENT, quote_component_value
from ....entities.trade_component import TradeComponent, DEFAULT_TRADE_COMPONENT, trade_component_value
from .params import DoubleExponentialMovingAverageLengthParams, DoubleExponentialMovingAverageSmoothingFactorParams


_EPSILON = 0.00000001


class DoubleExponentialMovingAverage(Indicator):
    """Computes the Double Exponential Moving Average (DEMA).

    DEMA = 2*EMA1 - EMA2, where EMA2 = EMA(EMA1).
    """

    def __init__(self, length: int, smoothing_factor: float,
                 length2: int, first_is_average: bool,
                 mnemonic: str, description: str,
                 bar_func, quote_func, trade_func) -> None:
        self._line = LineIndicator(mnemonic, description, bar_func, quote_func, trade_func, self.update)
        self._smoothing_factor: float = smoothing_factor
        self._length: int = length
        self._length2: int = length2
        self._first_is_average: bool = first_is_average
        self._sum: float = 0.0
        self._ema1: float = 0.0
        self._ema2: float = 0.0
        self._count: int = 0
        self._primed: bool = False

    @staticmethod
    def from_length(params: DoubleExponentialMovingAverageLengthParams) -> 'DoubleExponentialMovingAverage':
        """Creates a DEMA from length-based parameters."""
        return _new_dema(params.length, math.nan, params.first_is_average,
                         params.bar_component, params.quote_component, params.trade_component)

    @staticmethod
    def from_smoothing_factor(params: DoubleExponentialMovingAverageSmoothingFactorParams) -> 'DoubleExponentialMovingAverage':
        """Creates a DEMA from smoothing-factor-based parameters."""
        return _new_dema(0, params.smoothing_factor, params.first_is_average,
                         params.bar_component, params.quote_component, params.trade_component)

    def is_primed(self) -> bool:
        return self._primed

    def metadata(self) -> Metadata:
        return build_metadata(
            Identifier.DOUBLE_EXPONENTIAL_MOVING_AVERAGE,
            self._line.mnemonic,
            self._line.description,
            [OutputText(self._line.mnemonic, self._line.description)],
        )

    def update(self, sample: float) -> float:
        if math.isnan(sample):
            return sample

        if self._primed:
            sf = self._smoothing_factor
            v1 = self._ema1
            v2 = self._ema2
            v1 += (sample - v1) * sf
            v2 += (v1 - v2) * sf
            self._ema1 = v1
            self._ema2 = v2
            return 2.0 * v1 - v2

        self._count += 1
        if self._first_is_average:
            if self._count == 1:
                self._sum = sample
            elif self._length >= self._count:
                self._sum += sample
                if self._length == self._count:
                    self._ema1 = self._sum / self._length
                    self._sum = self._ema1
            else:
                self._ema1 += (sample - self._ema1) * self._smoothing_factor
                self._sum += self._ema1
                if self._length2 == self._count:
                    self._primed = True
                    self._ema2 = self._sum / self._length
                    return 2.0 * self._ema1 - self._ema2
        else:  # Metastock
            if self._count == 1:
                self._ema1 = sample
            elif self._length >= self._count:
                self._ema1 += (sample - self._ema1) * self._smoothing_factor
                if self._length == self._count:
                    self._ema2 = self._ema1
            else:
                self._ema1 += (sample - self._ema1) * self._smoothing_factor
                self._ema2 += (self._ema1 - self._ema2) * self._smoothing_factor
                if self._length2 == self._count:
                    self._primed = True
                    return 2.0 * self._ema1 - self._ema2

        return math.nan

    def update_scalar(self, sample: Scalar) -> Output:
        return self._line.update_scalar(sample)

    def update_bar(self, sample: Bar) -> Output:
        return self._line.update_bar(sample)

    def update_quote(self, sample: Quote) -> Output:
        return self._line.update_quote(sample)

    def update_trade(self, sample: Trade) -> Output:
        return self._line.update_trade(sample)


def _new_dema(length: int, alpha: float, first_is_average: bool,
              bc_param, qc_param, tc_param) -> DoubleExponentialMovingAverage:
    """Internal constructor for DEMA."""
    invalid = "invalid double exponential moving average parameters"

    bc = bc_param if bc_param is not None else DEFAULT_BAR_COMPONENT
    qc = qc_param if qc_param is not None else DEFAULT_QUOTE_COMPONENT
    tc = tc_param if tc_param is not None else DEFAULT_TRADE_COMPONENT

    if math.isnan(alpha):
        if length < 1:
            raise ValueError(f"{invalid}: length should be positive")
        alpha = 2.0 / (1 + length)
        mnemonic = f"dema({length}{component_triple_mnemonic(bc, qc, tc)})"
    else:
        if alpha < 0.0 or alpha > 1.0:
            raise ValueError(f"{invalid}: smoothing factor should be in range [0, 1]")
        if alpha < _EPSILON:
            alpha = _EPSILON
        length = int(round(2.0 / alpha)) - 1
        mnemonic = f"dema({length}, {alpha:.8f}{component_triple_mnemonic(bc, qc, tc)})"

    bar_func = bar_component_value(bc)
    quote_func = quote_component_value(qc)
    trade_func = trade_component_value(tc)

    length2 = 2 * length - 1
    description = f"Double exponential moving average {mnemonic}"

    return DoubleExponentialMovingAverage(length, alpha, length2, first_is_average,
                                          mnemonic, description,
                                          bar_func, quote_func, trade_func)
